src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -44,18 +40,28 @@ internal class DefaultProjectSnapshotManager : ProjectSnapshotManagerBase
 
     public DefaultProjectSnapshotManager(
         IErrorReporter errorReporter,
-        IEnumerable<ProjectSnapshotChangeTrigger> triggers,
+        IEnumerable<IProjectSnapshotChangeTrigger> triggers,
         Workspace workspace)
     {
-        _triggers = triggers?.OrderByDescending(trigger => trigger.InitializePriority).ToArray() ?? throw new ArgumentNullException(nameof(triggers));
         Workspace = workspace ?? throw new ArgumentNullException(nameof(workspace));
         ErrorReporter = errorReporter ?? throw new ArgumentNullException(nameof(errorReporter));
 
         using (_rwLocker.EnterReadLock())
         {
-            for (var i = 0; i < _triggers.Length; i++)
+            foreach (var trigger in triggers)
```

Comment:
We're in a read lock so we won't be able to trigger a changed event anyways (they all need write locks). Does that need to be different?

Reply:
I don't know, you added the lock 😛

This doesn't change the semantics of anything. Previously it looped through one collection that was ordered, and initialized them under a lock. Now it loops through twice, initializing them under a lock. I don't know why we would need to initialize under a lock, but I'm assuming you found a case where it was needed, so left it as is

Reply:
I added because there was a thread assert before, and this was the only place that wasn't just initializing the fields. Probably can be removed tbh, I was just pointing out that idk why priority would matter if they are all initialized before they could get a change notification

Reply:
> I was just pointing out that idk why priority would matter if they are all initialized before they could get a change notification

Ohhh yes! Completely missed that point, but its a very good one. Let me dig a little deeper into the one priority trigger and see if it holds up.

Reply:
I've thought about this more, and I think this still needs to be here. You're right about no notifications can fire while things are being initialized, but the eventual notifications will still happen in initialization order, and the high priority service needs to be the one to get those changes first, before other services. ie, when you open a document, we need it to be marked open by the high priority service, before the open document generator gets a notification and thinks its not open, so no-ops.

Given this is only used once, and only for open documents, I think a better long term approach here would be to have explicit document open/close `ProjectChangeKind`s, rather than the one `DocumentChanged` we currently have, but thats a bit more than I want to bite off right now. Also its probably all moot when we move to source generators 😁